muxink/src/demux.rs

```diff
@@ -0,0 +1,254 @@
+//! Stream demultiplexing
+//!
+//! Demultiplexes a Stream of Bytes into multiple channels. Up to 256 channels are supported, and
+//! if messages are present on a channel but there isn't an associated DemultiplexerHandle for that
+//! channel, then the Stream will never poll as Ready.
+
+use std::{
+    pin::Pin,
+    sync::{Arc, Mutex},
+    task::{Context, Poll},
+};
+
+use bytes::Bytes;
+use futures::{stream::Fuse, Stream, StreamExt};
+
+/// A frame demultiplexer.
+///
+/// A demultiplexer is not used directly, but used to spawn demultiplexing handles.
+///
+/// TODO What if someone sends data to a channel for which there is no handle?
+///      I can think of two reasonable responses:
+///         1. return an error to the handle which saw this message.
+///         2. drop all messages we receive which don't have a corresponding `DemultiplexerHandle`
+///            yet.
+///         3. allow messages to sit forever and block the rest of the handles, preferring whoever
+///            is sending us the messages to filter out ones which aren't for a channel we're
+///            listening on. this is already what happens if a `DemultiplexerHandle` for any
+///            channel which has messages in the stream doesn't ever take them out.
+pub struct Demultiplexer<S> {
+    stream: Fuse<S>,
+    next_frame: Option<(u8, Bytes)>,
+}
+
+impl<S: Stream> Demultiplexer<S> {
+    /// Creates a new demultiplexer with the given underlying stream.
+    pub fn new(stream: S) -> Demultiplexer<S> {
+        Demultiplexer {
+            stream: stream.fuse(),
+            next_frame: None,
+        }
+    }
+}
+
+impl<S> Demultiplexer<S> {
+    /// Creates a handle listening for frames on the given channel.
+    ///
+    /// Any item on this channel sent to the `Stream` underlying the `Demultiplexer` we used to
+    /// create this handle will be read only when all other messages for other channels have been
```

Comment:
Is this slightly wrong? I would put it differently, e.g. "Items received through a given handle may be blocked if other handles on the same `Demultiplexer` are not polled at the same time."